restore pinpointing for CheckMaxTime

eriknordmark opened the pull request:
This can hopefully be handled in a generic event loop logic.

Note that without this the CheckMaxTime provides the same function and line number for all the call sites hence we can't tell which handler was slow.

kalyan-nidumolu commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -396,3 +396,24 @@ func CheckMaxTime(agentName string, start time.Time) {
+	if agentName == "nim" {
+		errTime = errorTimeNim
+		warnTime = warningTimeNim
+	}
```
Why is there nim hardcoding in agent.log?? Can we make it more modular? Let the agents pass error / warning times ??

eriknordmark replied:
This code didn't change.
Note that I need two copies of the two flavors of the function since it uses the caller.

deitch replied:
> Note that I need two copies of the two flavors of the function since it uses the caller

Yup, makes sense.

deitch replied:
I agree with @kalyan-nidumolu ; we should remove the hard-coded custom timing. If we need a special case, we can have a map with them, and it can look up the times in a map with a default. We probably should avoid hardcoding  a specific agent in here.

eriknordmark replied:
That's fine, however this PR is to fix a regression caused a recent PR: https://github.com/lf-edge/eve/pull/440
I can instead back out that PR if you prefer. But I think you have more important things on your plate.

deitch replied:
Yeah fair enough.

eriknordmark replied:
I've removed this and moved the values out into the agents.

deitch commented:
Working on phone. Will get to it in a few hours.

deitch commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -396,3 +396,24 @@ func CheckMaxTime(agentName string, start time.Time) {
+	}
+	elapsed := time.Since(start)
+	if elapsed > errTime {
+		_, fn, line, _ := runtime.Caller(1)
```
Is the purpose of `CheckMaxTimeTopic` not to avoid the whole `runtime.Caller(1)` magic, and instead by purely functional, i.e. "I am a function that will check the time you passed, and if it is a problem, will log a message with the message parameters you gave me"?

eriknordmark replied:
Issue is that we have several call sites from different select loops and we might want to know which one caused the issue. Alternatively we handle that by making sure we have log.Info messages each time we move from one select loop.

deitch replied:
Understood. It does make it non-functional, which can trip people up (as it did me). More relevantly, though, is that line numbers are useful for crashes, that are resolved by developers; not so much for performance issues, where you care what caused X to spend so much time, which is more operational.

eriknordmark replied:
FWIW I removed this use of Caller by making pubsub and other callers specify the topic aka collection which had the time issue.

deitch commented on pkg/pillar/agentlog/agentlog.go:
```diff
@@ -396,3 +396,24 @@ func CheckMaxTime(agentName string, start time.Time) {
+	elapsed := time.Since(start)
+	if elapsed > errTime {
+		_, fn, line, _ := runtime.Caller(1)
+		log.Errorf("%s handler at %s:%d in %s XXX took a long time: %d",
```
I think we can avoid redundancy by creating the message outside of the `if` statement, and just do:

```go
elapsed := time.Since(start)
msg := fmt.Sprintf("%s handler in %s XXX took a long time: %d", topic, agentName, elapsed/time.Second)
if elapsed > errTime {
    log.Error(msg)
} else if elapsed > warnTime {
// etc.
```

eriknordmark commented:
Note that this can be encapsulated inside pubsub.Subscription; can set the time limits as part of the Subscription object and the ProcessChange itself can do the checks.
But for the time being let's keep this external to pubsub; note that we have some  non-subscription handlers such as timer-driven routines which we should also check for long runtimes.

eriknordmark commented:
@deitch @kalyan-nidumolu I've bit the bullet and moved the checks inside pubsub and removed the code which knew about nim being special. Please take a look.

deitch commented:
I think I see what you did. You moved the actual time checking into `ProcessChange()` itself, since everything calls that? And then save the expected limits on each topic's subscription?

I don't love adding yet another 2 elements to `Subscription` as I am looking to restructure it, but that is more my problem. Overall, I like it. It eliminates a lot of boilerplate and moves the logic into one place.

eriknordmark commented:
> I don't love adding yet another 2 elements to `Subscription` as I am looking to restructure it, but that is more my problem. Overall, I like it. It eliminates a lot of boilerplate and moves the logic into one place.

Yes, if we could have some form of global init of the pubsub module where each agent can provide arguments it would be less duplicate code. Perhaps we should introduce a pubsubContext which can be initialized; don't know how that will fit with your refactoring. So food for thought for now.
